#include "diem_input.h"
#include "ui_diem_input.h"
#include "QHBoxLayout"
#include "DS_SinhVien.h"


#include <QMessageBox>
#include <QString>
#include <QFile>
#include <QTextStream>
#include <fstream>
#include <sstream>
#include <QDebug>
#include <QWidget>
#include <QDir>

using namespace std;


diem_input::diem_input(QWidget *parent,const QString &a)
    : QWidget(parent),ui(new Ui::diem_input)
{
    ui->setupUi(this);
    danhSachNhapDiem = nullptr;// t·∫°o ds r·ªóng
    updateTable();//c·∫≠p nh·∫≠t b·∫£ng

}


diem_input::~diem_input()
{
    delete ui;
}
// -------------------- T√¨m sinh vi√™n  --------------------
SinhVien* diem_input::timSinhVien(QString maSV) {
    QFile file("sinhvien.txt");
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) return nullptr;

    QTextStream in(&file);
    while(!in.atEnd()) {
        QStringList t = in.readLine().split("|");
        if(t.size()<5) continue;
        if(t[1].trimmed() == maSV.trimmed()) {
            SinhVien* sv = new SinhVien;
            strcpy(sv->MASV, t[1].toUtf8().data());
            strcpy(sv->HO, t[2].toUtf8().data());
            strcpy(sv->TEN, t[3].toUtf8().data());
            return sv;
        }
    }
    return nullptr;
}

// -------------------- T√¨m m√£ l·ªõp t√≠n ch·ªâ --------------------
int diem_input::timMaLopTC(QString tenMH, QString nienKhoa, int hocKy, int nhom)
{
    QFile file("LopTinChi.txt");
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        return -1;

    QTextStream in(&file);
    while (!in.atEnd()) {
        QStringList t = in.readLine().split("|");
        if (t.size() < 6) continue;
        if (t[2] == tenMH &&
            t[3] == nienKhoa &&
            t[4].toInt() == hocKy &&
            t[5].toInt() == nhom)

            return t[0].toInt();// tr·∫£ v·ªÅ m√£ l·ªõp
    }
    return -1;
}

// -------------------- T√¨m con tr·ªè l·ªõp t√≠n ch·ªâ t·ª´ m√£ l·ªõp --------------------
LopTinChi* diem_input::timLopTinChi(int maLop)
{
    QFile file("LopTinChi.txt");
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        return nullptr;

    QTextStream in(&file);
    while (!in.atEnd()) {
        QStringList t = in.readLine().split("|");
        if (t.size() < 6) continue;

        if (t[0].toInt() == maLop) {
            LopTinChi* lop = new LopTinChi;
            lop->MALOPTC = t[0].toInt();
            lop->MAMH = t[1];
            lop->TENMH = t[2];
            lop->NienKhoa = t[3];
            lop->Hocky = t[4].toInt();
            lop->Nhom = t[5].toInt();
            return lop;
        }
    }
    return nullptr;
}

// -------------------- T·∫°o danh s√°ch nh·∫≠p ƒëi·ªÉm --------------------
PTRBangNhapDiem diem_input::taoDanhSachNhapDiem(int maloptc)
{
    QFile file("danhsachdangky.txt");
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        return nullptr;

    QTextStream in(&file);
    PTRBangNhapDiem head = nullptr, tail = nullptr;
    int stt = 1;

    while (!in.atEnd()) {
        QString line = in.readLine();
        QStringList t = line.split("|");
        if (t.size() < 4) continue;

        int maLop = t[0].toInt(); // m√£ l·ªõp t·ª´ file
        QString maSV = t[1];
        float diem = t[2].toFloat(); // ƒëi·ªÉm c√≥ th·ªÉ l√† -1, 0, >0

        if (maLop != maloptc) continue;

        SinhVien* sv = timSinhVien(maSV);
        if (!sv) {
            qDebug() << "Khong tim thay sinh vien:" << maSV;
            continue;
        }

        PTRBangNhapDiem node = new NodeBangNhapDiem;
        node->data.STT = stt++;
        node->data.controSV = sv;               // con tr·ªè sinh vi√™n
        node->data.controMaLTC = timLopTinChi(maLop); // con tr·ªè l·ªõp t√≠n ch·ªâ
        node->data.DIEM_INPUT = diem;           // gi·ªØ -1, 0 ho·∫∑c ƒëi·ªÉm th·ª±c
        node->next = nullptr;

        if (!head) head = tail = node;
        else { tail->next = node; tail = node; }
    }

    if (!head)
        qDebug() << "Khong co sinh vien nao trong maloptc:" << maloptc;

    return head;
}


//l∆∞u ƒëi·ªÉm
void diem_input::luuDiem() {
    QFile file("danhsachdangky.txt");
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        QMessageBox::warning(this, "L·ªói", "Kh√¥ng th·ªÉ m·ªü file ƒë·ªÉ ƒë·ªçc!");
        return;
    }

    // ƒë·ªçc file c≈© v√†o danh s√°ch
    QStringList lines;
    QTextStream in(&file);
    while (!in.atEnd()) {
        lines << in.readLine();
    }
    file.close();

    // c·∫≠p nh·∫≠t ƒëi·ªÉm
    for (int i = 0; i < lines.size(); ++i) {
        QStringList t = lines[i].split("|");
        if (t.size() < 4) continue;

        int maLop = t[0].toInt();
        QString maSV = t[1];
        QString trangThaiDangKy = t[3];  // üîí gi·ªØ nguy√™n

        NodeBangNhapDiem* p = danhSachNhapDiem;
        while (p) {
            if (p->data.controMaLTC->MALOPTC == maLop &&
                QString::fromUtf8(p->data.controSV->MASV) == maSV) {

                float diem = p->data.DIEM_INPUT;  // -1 ho·∫∑c 0-10

                lines[i] = QString("%1|%2|%3|%4")
                               .arg(maLop)
                               .arg(maSV)
                               .arg(diem)
                               .arg(trangThaiDangKy); //  KH√îNG ƒê·ªîI
                break;
            }
            p = p->next;
        }
    }
    // ghi l·∫°i file
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        QMessageBox::warning(this, "L·ªói", "Kh√¥ng th·ªÉ m·ªü file ƒë·ªÉ l∆∞u!");
        return;
    }
    QTextStream out(&file);
    for (const QString &line : lines) {
        out << line << "\n";
    }
    file.close();

    QMessageBox::information(this, "L∆∞u ƒëi·ªÉm", "ƒê√£ l∆∞u danh s√°ch ƒëi·ªÉm m·ªõi!");
}




//giao di·ªán

void diem_input::on_pushButton_TraBangDiem_clicked()
{// l·∫•y in4 t·ª´ lineEdit
    QString tenMH = ui->lineEdit__TenMon->text().trimmed();
    QString nienKhoa = ui->lineEdit__NienKhoa->text().trimmed();
    int hocKy = ui->lineEdit__Hocky->text().toInt();
    int nhom = ui->lineEdit__Nhom->text().toInt();

    int maloptc = timMaLopTC(tenMH, nienKhoa, hocKy, nhom);

    if(tenMH.isEmpty()) {
        QMessageBox::warning(this,"L·ªói","Ch∆∞a nh·∫≠p t√™n m√¥n");
        return;
    }
    if(nienKhoa.isEmpty()) {
        QMessageBox::warning(this,"L·ªói","Ch∆∞a nh·∫≠p ni√™n kh√≥a");
        return;
    }
    if(hocKy<= 0) {
        QMessageBox::warning(this,"L·ªói","Ch∆∞a nh·∫≠p h·ªçc k·ª≥");
        return;
    }
    if(nhom<= 0) {
        QMessageBox::warning(this,"L·ªói","Ch∆∞a nh·∫≠p nh√≥m");
        return;
    }
    if (maloptc == -1) {
        QMessageBox::warning(this, "Loi", "Khong tim thay lop tin chi");
        return;
    }

    danhSachNhapDiem = taoDanhSachNhapDiem(maloptc);
    if (!danhSachNhapDiem) {
        QMessageBox::warning(this, "L·ªói", "Kh√¥ng c√≥ sinh vi√™n ƒë·ªÉ hi·ªÉn th·ªã");
        return;
    }
    updateTable();
}

void diem_input::updateTable() {
    ui->BangDiem_Widget->verticalHeader()->setVisible(false);  // x√≥a stt mƒë
    ui->BangDiem_Widget->setColumnCount(6);
    QStringList tieude = {"STT", "M√£ SV", "H·ªç","T√™n", "ƒêi·ªÉm", "Ch·ª©c nƒÉng"};
    ui->BangDiem_Widget->setHorizontalHeaderLabels(tieude);
    ui->BangDiem_Widget->horizontalHeader()->setSectionResizeMode(QHeaderView::Stretch);

    ui->BangDiem_Widget->setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
    ui->BangDiem_Widget->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);

    ui->BangDiem_Widget->clearContents();
    ui->BangDiem_Widget->setRowCount(0);


    NodeBangNhapDiem *p = danhSachNhapDiem;
    int row = 0;

    while (p) {
        ui->BangDiem_Widget->insertRow(row);// ch√®n h√†ng
        ui->BangDiem_Widget->setItem(row, 0, new QTableWidgetItem(QString::number(p->data.STT)));
        ui->BangDiem_Widget->item(row, 0)->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled);

        ui->BangDiem_Widget->setItem(row, 1, new QTableWidgetItem(QString::fromUtf8(p->data.controSV->MASV)));
        ui->BangDiem_Widget->item(row, 1)->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled);

        ui->BangDiem_Widget->setItem(row, 2, new QTableWidgetItem(QString::fromUtf8(p->data.controSV->HO)));
        ui->BangDiem_Widget->item(row, 2)->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled);

        ui->BangDiem_Widget->setItem(row, 3, new QTableWidgetItem(QString::fromUtf8(p->data.controSV->TEN)));
        ui->BangDiem_Widget->item(row, 3)->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled);

        QString diemStr = (p->data.DIEM_INPUT < 0.0f) ? "ch∆∞a c√≥": QString::number(p->data.DIEM_INPUT);
        ui->BangDiem_Widget->setItem(row, 4, new QTableWidgetItem(diemStr));
        ui->BangDiem_Widget->item(row, 4)->setFlags(Qt::ItemIsSelectable | Qt::ItemIsEnabled | Qt::ItemIsEditable);       // cho ph√©p ch·ªânh s·ª≠a tr·ª±c ti·∫øp c·ªôt ƒëi·ªÉm

        // t·∫°o n√∫t hi·ªáu ch·ªânh
        QPushButton* btnNhap = new QPushButton("Nh·∫≠ppp");
        QPushButton* btnXoa = new QPushButton("X√≥a");
        QWidget* btnWidget = new QWidget;
        QHBoxLayout* layout = new QHBoxLayout(btnWidget);
        layout->addWidget(btnNhap);
        layout->addWidget(btnXoa);
        layout->setContentsMargins(2, 0, 2, 0);
        ui->BangDiem_Widget->setCellWidget(row, 5, btnWidget);

        //int currentRow = row;
        connect(btnNhap, &QPushButton::clicked, this, [=]() { onNhapDiemClicked(row); });
        connect(btnXoa, &QPushButton::clicked, this, [=]() { onXoaDiemClicked(row); });

        p = p->next;
        row++;
    }
    ui->BangDiem_Widget->viewport()->update(); // refresh table
}
void diem_input::onNhapDiemClicked(int row) {
    QString maSV = ui->BangDiem_Widget->item(row, 1)->text(); // l·∫•y in4 t·ª´ √¥ m√£ sv
    QString diemStr = ui->BangDiem_Widget->item(row, 4)->text(); // l·∫•y in4 t·ª´ √¥ ƒëi·ªÉm

    // nh·∫≠n bi·∫øt ng d√πng nh·∫≠p s·ªë hay ch·ªØ
    bool ok;
    float diem = diemStr.toFloat(&ok);

    if (!ok) {  // n·∫øu kh√¥ng ph·∫£i s·ªë
        QMessageBox::warning(this, "L·ªói", "ƒêi·ªÉm ph·∫£i l√† s·ªë t·ª´ 0 ƒë·∫øn 10!");
        return;
    }

    if (diem < 0.0f || diem > 10.0f) { // ki·ªÉm tra kho·∫£ng
        QMessageBox::warning(this, "L·ªói", "ƒêi·ªÉm ph·∫£i t·ª´ 0 ƒë·∫øn 10!");
        return;
    }

    NodeBangNhapDiem* p = danhSachNhapDiem;
    while(p) {
        if(QString::fromUtf8(p->data.controSV->MASV) == maSV){
            p->data.DIEM_INPUT = diem;
            break;
        }
        p = p->next;
    }
    QString masv = ui->BangDiem_Widget->item(row, 1)->text(); // gi·∫£ s·ª≠ c·ªôt 0 l√† m√£ SV
    QMessageBox::information(this,
                             "Nhap diem",
                             "Nhap diem cho " + masv + ": " + QString::number(diem));
    luuDiem();
    updateTable();
}

void diem_input::onXoaDiemClicked(int row) {
    QString maSV = ui->BangDiem_Widget->item(row, 1)->text();
    NodeBangNhapDiem* p = danhSachNhapDiem;
    while(p){
        if(QString::fromUtf8(p->data.controSV->MASV) == maSV){
            p->data.DIEM_INPUT = -1;
            break;
        }
        p = p->next;
    }
    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, "X√°c nh·∫≠n x√≥a",
                                  "X√≥a ƒëi·ªÉm c·ªßa SV " + maSV + "?",
                                  QMessageBox::Yes | QMessageBox::No);
    if (reply == QMessageBox::Yes) {
    ui->BangDiem_Widget->item(row, 4)->setText("ch∆∞a c√≥"); // ch·ªâ c·∫≠p nh·∫≠t l·∫°i c·ªôt ƒëi·ªÉm trong table
    luuDiem();
    updateTable();
    }
}




